test/functional/lib/tests.cpp

```diff
@@ -6487,6 +6620,82 @@ GenericTxChecksumOffloadIp()
     MpTxFlush(GenericMp);
 }
 
+VOID
+GenericRxChecksumOffloadIp() {
+    const BOOLEAN Rx = TRUE, Tx = FALSE;
+    auto If = FnMpIf;
+    UINT16 LocalPort, RemotePort;
+    ETHERNET_ADDRESS LocalHw, RemoteHw;
+    INET_ADDR LocalIp, RemoteIp;
+    auto Xsk = CreateAndBindSocket(If.GetIfIndex(), If.GetQueueId(), Rx, Tx, XDP_GENERIC);
+    auto UdpSocket = CreateUdpSocket(AF_INET, NULL, &LocalPort);
+    auto GenericMp = MpOpenGeneric(If.GetIfIndex());
+
+
+    // Routine Description:
+    //     This test verifies the metadata that indicates RX checksum offloads were done for IP.
+    //     It does so by injecting a UDP frame with a valid IP checksum and verifying that
+    //     the metadata is correctly set by the miniport driver.
+
+    EnableRxChecksumOffload(&Xsk);
+    ActivateSocket(&Xsk, Rx, Tx);
+
+    Xsk.RxProgram =
+        SocketAttachRxProgram(If.GetIfIndex(), &XdpInspectRxL2, If.GetQueueId(), XDP_GENERIC, Xsk.Handle.get());
+
+    LocalPort = htons(1234);
+    RemotePort = htons(4321);
+    If.GetHwAddress(&LocalHw);
+    If.GetRemoteHwAddress(&RemoteHw);
+    If.GetIpv4Address(&LocalIp.Ipv4);
+    If.GetRemoteIpv4Address(&RemoteIp.Ipv4);
+
+    // Construct a valid IPv4 + UDP frame with a valid IP checksum
+    UCHAR UdpPayload[] = "GenericRxChecksumOffloadIp";
+    UCHAR UdpFrame[UDP_HEADER_STORAGE + sizeof(UdpPayload)];
+    UINT32 UdpFrameLength = sizeof(UdpFrame);
+
+    TEST_TRUE(
+        PktBuildUdpFrame(
+            UdpFrame, &UdpFrameLength, UdpPayload, sizeof(UdpPayload), &LocalHw,
+            &RemoteHw, AF_INET, &LocalIp, &RemoteIp, LocalPort, RemotePort));
+
+    RX_FRAME RxFrame;
+    RxInitializeFrame(&RxFrame, If.GetQueueId(), UdpFrame, UdpFrameLength);
+    RxFrameSetChecksumOffloadState(&RxFrame, FALSE, FALSE, FALSE, FALSE, TRUE, FALSE);
+
+    // Inject the frame as if it came from the wire
+    TEST_HRESULT(MpRxEnqueueFrame(GenericMp, &RxFrame));
+
+    //
+    // Produce one XSK fill descriptor.
+    //
+    SocketProduceRxFill(&Xsk, 1);
+
+    // Flush RX path to make frame visible to the socket
+    MpRxFlush(GenericMp);
+
+    UINT32 ConsumerIndex;
+    TEST_EQUAL(1, XskRingConsumerReserve(&Xsk.Rings.Rx, 1, &ConsumerIndex));
+
+    printf("ConsumerIndex: %u\n", ConsumerIndex);
+
+    XSK_FRAME_DESCRIPTOR *RxDesc = SocketGetRxFrameDesc(&Xsk, ConsumerIndex++);
+    TEST_TRUE(Xsk.Extensions.RxFrameChecksumExtension != 0);
+
+    // Get and validate checksum metadata
+    XDP_FRAME_CHECKSUM *Checksum =
+        (XDP_FRAME_CHECKSUM *)RTL_PTR_ADD(RxDesc, Xsk.Extensions.RxFrameChecksumExtension);
+
+    printf("Xsk.Extensions.RxFrameChecksumExtension: %u\n", Xsk.Extensions.RxFrameChecksumExtension);
+
+    TEST_EQUAL(XdpFrameRxChecksumEvaluationSucceeded, Checksum->Layer3);
+    TEST_EQUAL(XdpFrameRxChecksumEvaluationNotChecked, Checksum->Layer4);
+
+    XskRingConsumerRelease(&Xsk.Rings.Rx, 1);
+}
+
+
 VOID
 GenericTxChecksumOffloadTcp(
     ADDRESS_FAMILY Af
```

Comment:
@mtfriesen why do we create a udpsocket in this TCP test function?

Reply:
Good question - looks like a bug.

Reply:
Fixed in #862